controllers/operatorpolicy_controller.go

```diff
@@ -60,166 +70,211 @@ var _ reconcile.Reconciler = &OperatorPolicyReconciler{}
-func (r *OperatorPolicyReconciler) handleSinglePolicy(
+// createPolicyResources encapsulates the logic for creating resources specified within an operator policy.
+// This should normally happen when the policy is initially applied to the cluster. Creates an OperatorGroup
+// if specified, otherwise defaults to using an allnamespaces OperatorGroup.
+func (r *OperatorPolicyReconciler) createPolicyResources(
 	policy *policyv1beta1.OperatorPolicy,
-) error {
-	OpLog.Info("Handling OperatorPolicy", "policy", policy.Name)
+	watcher depclient.ObjectIdentifier,
+) (ctrl.Result, error) {
+	// Retrieve og from cache
+	ogNamespace := policy.Spec.Subscription.Namespace
+	if policy.Spec.OperatorGroup != nil {
+		ogNamespace = policy.Spec.OperatorGroup.Namespace
+	}
 
-	subscriptionSpec := new(operatorv1alpha1.Subscription)
-	err := r.Get(context.TODO(),
-		types.NamespacedName{Namespace: subscriptionSpec.Namespace, Name: subscriptionSpec.Name},
-		subscriptionSpec)
-	exists := !errors.IsNotFound(err)
-	shouldExist := strings.EqualFold(string(policy.Spec.ComplianceType), string(policyv1.MustHave))
+	ogList, err := r.DynamicWatcher.ListFromCache(
+		operatorgroupgvk,
+		ogNamespace, nil)
+	if err != nil {
+		panic(err)
+	}
 
-	// Object does not exist but it should exist, create object
-	if !exists && shouldExist {
+	//  Create og, then trigger reconcile
+	if len(ogList) == 0 {
 		if strings.EqualFold(string(policy.Spec.RemediationAction), string(policyv1.Enforce)) {
-			OpLog.Info("creating kind " + subscriptionSpec.Kind + " in ns " + subscriptionSpec.Namespace)
-			subscriptionSpec := buildSubscription(policy, subscriptionSpec)
-			err = r.Create(context.TODO(), subscriptionSpec)
+			ogSpec := buildOperatorGroup(policy)
 
+			err = r.Create(context.TODO(), ogSpec)
 			if err != nil {
+				OpLog.Error(err, "Error while creating OperatorGroup")
 				r.setCompliance(policy, policyv1.NonCompliant)
-				OpLog.Error(err, "Could not handle missing musthave object")
 
-				return err
+				err = r.DynamicWatcher.EndQueryBatch(watcher)
+				if err != nil {
+					panic(err)
+				}
+
+				return reconcile.Result{}, err
 			}
 
-			// Currently creates an OperatorGroup for every Subscription
-			// in the same ns, and defaults to targeting all ns.
-			// Future implementations will enable targeting ns based on
-			// installModes supported by the CSV. Also, only one OperatorGroup
-			// should exist in each ns
-			operatorGroup := buildOperatorGroup(policy)
-			err = r.Create(context.TODO(), operatorGroup)
+			// Created successfully, requeue result
+			r.setCompliance(policy, policyv1.NonCompliant)
 
+			err = r.DynamicWatcher.EndQueryBatch(watcher)
 			if err != nil {
-				r.setCompliance(policy, policyv1.NonCompliant)
-				OpLog.Error(err, "Could not handle missing musthave object")
-
-				return err
+				panic(err)
 			}
 
-			r.setCompliance(policy, policyv1.Compliant)
+			return reconcile.Result{Requeue: true}, err
```

Comment:
Why do you have to requeue in this case?

Reply:
So the pattern Justin, Jeffrey and I agreed upon is to create one resource at a time and return early (and requeue) so that each reconcile is more cleanly divided into different logical chunks and it's purpose is clear. For example: First reconcile is to create the OperatorGroup, second one for Subscription.

Prior to this, the controller would try to create all the resources in one reconcile.

Reply:
Could you explain why this pattern was chosen?

I'm thinking that it will not be efficient and it doesn't really add to be better readability/maintainability in my opinion. Why I say it's not efficient is that the first reconcile creates the operator group. Then after the requeue, it'll check to see if the operator group matches, and so on and so forth.

Reply:
I suggested this pattern. I had a dream of what I wanted it to look like, and I still believe the idea would help with modularization and readability, but I concede it would be less efficient than an alternative.

I was hoping the main reconcile could simplify down to something like this:
```golang
opGroup, ogNeedsAction := r.CheckOperatorGroup(policy)
if ogNeedsAction {
    return r.HandleOperatorGroup(policy, opGroup)
}

sub, subNeedsAction := r.CheckSubscription(policy)
if subNeedsAction {
    return r.HandleSubscription(policy, sub)
}

installPlan, ipNeedsAction := r.CheckInstallPlan(policy, sub)
if ipNeedsAction {
    return r.HandleInstallPlan(policy, installPlan, sub)
}
```

If each `Check` function could be read-only API calls, and therefore be likely to just use the cache, they would be quite fast/cheap and I wouldn't mind running them potentially many times due to repeated reconciles.

Then the `Handle` functions would have some write step, creating or updating a resource, possibly just the updating policy's status. I feel like there are a lot of different possible outcomes of the `Handle` functions, with some outcomes blocking future work, and figuring out how to handle all of the outcomes would become a maze of logic. I feel it is simpler to just restart at that point, and if the actions were correctly taken (and have "stuck" on the cluster) than the next corresponding `Check` will report that no further action is needed.

Additionally, since the `Handle` functions do some API calls, they would take more time, and things fetched from previous steps might no longer be up to date. Bugs, or at least possible update errors, could then appear. Restarting, and just doing "fast" things until another action needs to be taken might avoid those.

Reply:
I think you'll wind up having to check everything anyways in the inform case since you probably want to provide an exhaustive list of the non-compliant "related objects", so then the code flows would be very different for inform/enforce.

I'm also not quite seeing the complexity that this simplifies. I just find it to be an unusual pattern and don't think it'd be hard to write this without the requeuing.

If you want to handle that in a separate PR to unblock this one, that'd be fine.